Instructions to run integration tests against quickstart

suneet-s opened the pull request:
This change makes it easier to run integration tests against a quickstart cluster running on a local development machine.

The primary motivation for this is debuggability, as you can have a local quickstart environment set up with remote debugging so you can easily put breakpoints in the cluster under test.

The other advantage of using this mode is that you can run an integration test against an older version of quickstart to verify that the test fails for older versions, but pass when the fix is applied in a newer version.

maytasm3 commented:
What is the motivation for running tests Using A Quickstart Cluster?

suneet-s commented:
@maytasm3 Yes - debuggability and speed.

I was trying to debug the integration tests. Unfortunately the docker environment is spun up and down as part of the integration tests, so it's hard to debug that environment. It's also quicker for me to test against since I usually have quickstart of a few different versions of druid on my dev machine. This makes it easier for me to verify that the integration test I'm adding fails for a new version, but works with my fix.

Also spinning up the docker environment can take some time, so I found it faster to point my tests to an already running quickstart

ccaominh commented on integration-tests/README.md:
```diff
@@ -66,7 +66,40 @@ Integration tests can also be run with either Java 8 or Java 11 by adding -Djvm.
+When writing integration tests, it can be helpful to test against a quickstart
+cluster so that you can set up remote debugging with in your developer
+environment. This section walks you through setting up the integration tests
+so that it can run against a quickstart cluster running on your development
```
How about linking against the quickstart docs so that folks know what is meant by a "quickstart cluster"?

suneet-s replied:
Done

ccaominh commented on integration-tests/README.md:
```diff
@@ -66,7 +66,40 @@ Integration tests can also be run with either Java 8 or Java 11 by adding -Djvm.
+Make sure you have at least 6GB of memory available before you run the tests.
+
+The tests rely on files in the test/resources folder to exist under the path /resources,
+so create a symlink to make them available
```
Suggestion: Add `:` at the end

(similar comment for other occurrences below)

ccaominh replied:
Having to create a new directory as a child of root seems strange. Is there a way to make it local to the druid repo directory?

suneet-s replied:
Done

ccaominh commented on integration-tests/README.md:
````diff
@@ -66,7 +66,40 @@ Integration tests can also be run with either Java 8 or Java 11 by adding -Djvm.
+  export CONFIG_FILE=${DRUID_HOME}/integration-tests/quickstart-it.json
+```
+Note that quickstart does not run with ssl, so to trick the integration tests
+we specify the `*_tls_url` in the config to be the same as the http url
````
Suggestion: Add `.` at the end

suneet-s replied:
Done

maytasm3 commented:
@suneet-s I think this is super useful as you can easily have more control over Druid cluster such as setting up remote debugging. One problem I see is that the docker environment do some set up of its own that is needed for some integration tests. For example, some confs set and tls/crl files generated for security IT or inserting rows into druid_segments metadatastore for the query IT (and also aws s3 keys to pull those segments). These most likely will not be configured in quickstart cluster and can result in false failures. Maybe it's worth mentioning the potential problems in the README so user does not except full compatibility out of the box

suneet-s commented on integration-tests/README.md:
```diff
@@ -66,7 +66,40 @@ Integration tests can also be run with either Java 8 or Java 11 by adding -Djvm.
+so that it can run against a quickstart cluster running on your development
+machine.
+
+Note that not all features run by default on a quickstart cluster, so it may
```
@maytasm3 I was worried about what you mentioned as a top level comment on this PR too. I tried to talk about it here, but maybe the warning is not big and scary enough. Any suggestions on how to make it more prominent? Also I wonder if I should point to the docker setup scripts for the next person who might find a test that fails when run against quickstart.

ccaominh replied:
Other places in the docs use a blockquote like: https://druid.apache.org/docs/latest/querying/lookups.html

maytasm3 replied:
I feel the term "feature" might be a little misleading. Some of the setup on the docker cluster are just pre-inserted data into the metadata store or certain files place at certain locations. The feature itself like pulling segments from pathSpec in metadatastore is always there. I think as we add more and more IT and/or setup scripts, it will also be hard to knows which can work out of the box on quickstart vs. which can only work on docker cluster. I think it might help if we can list out which groups can work out of the box on quickstart vs. which can only work on docker cluster in the README. In the file with the group definition (TestNGGroup) we can then list out special setup needed (if needed) for those groups.

suneet-s replied:
👍 Added a note on the setup, but it's kinda hard to know which groups depend on what right now, maybe after using this for a few months the test group will be more complete. I added instructions to this file and the group so that other devs can add more tests to this list.

ccaominh commented on integration-tests/src/test/java/org/apache/druid/tests/TestNGGroup.java:
```diff
@@ -33,29 +33,57 @@
   public static final String S3_INGESTION = "s3-ingestion";
+
+  /**
+   * This group os not part of CI explicitly. It allows you to run all the tests that have been tested with
```
If you push another commit, there's a small typo here: group os -> group is

maytasm3 commented:
@suneet-s You should also exclude the new group QUICKSTART_COMPATIBLE from travis "other integration test"

ccaominh commented:
> @suneet-s You should also exclude the new group QUICKSTART_COMPATIBLE from travis "other integration test".
> I feel like that "other integration test" instead of doing exclude should just do include

The rationale for excluding was that it's better to accidentally run tests in multiple jobs rather than having a test run in no jobs, but we can adjust the approach if needed.

maytasm3 commented:
> > @suneet-s You should also exclude the new group QUICKSTART_COMPATIBLE from travis "other integration test".
> > I feel like that "other integration test" instead of doing exclude should just do include
> 
> The rationale for excluding was that it's better to accidentally run tests in multiple jobs rather than having a test run in no jobs, but we can adjust the approach if needed.

That's true but I think as we add more and more IT, especially those unique ones that needs special setup, that exclude list will be super long and ugly

suneet-s commented:
> @suneet-s You should also exclude the new group QUICKSTART_COMPATIBLE from travis "other integration test".
> I feel like that "other integration test" instead of doing exclude should just do include

Actually, I think this might be wrong. The test is assigned to 2 groups. So if I exclude 1/2 groups, I think it might exclude the test. Need to check locally if that's the case. So much TestNG magic I'm learning about in this PR

suneet-s commented:
> > @suneet-s You should also exclude the new group QUICKSTART_COMPATIBLE from travis "other integration test".
> > I feel like that "other integration test" instead of doing exclude should just do include
> 
> Actually, I think this might be wrong. The test is assigned to 2 groups. So if I exclude 1/2 groups, I think it might exclude the test. Need to check locally if that's the case. So much TestNG magic I'm learning about in this PR

@maytasm3 I tried this locally, and it appears to run the test.

`mvn clean verify -P int-tests-config-file -Dit.test=ITCompactionTaskTest -DexcludedGroups=quickstart-compatible`

I'll look at the results of this travis run to make sure that ITCompactionTaskTest is actually run in this change

suneet-s commented:
The `ITCompactionTaskTest` was excluded in the other integration test run - https://travis-ci.org/github/apache/druid/jobs/667053767?utm_medium=notification&utm_source=github_status, so I reverted the change that explicitly excludes that test group and added more javadocs.

The `ITIndexerTest` was run coorectly - https://travis-ci.org/github/apache/druid/jobs/667053761?utm_medium=notification&utm_source=github_status

maytasm3 commented:
Does this means that ITIndexerTest will get run twice? (once in batch index and once more in other)

suneet-s commented:
I don't see it in the list of tests run by the "other" test job. I think it's excluded because of this in the config `-DexcludedGroups=batch-index`
